Step4 : 경로 조회 기능

impati opened the pull request:
안녕하세요 리뷰어님.! 

경로 조회 기능 구현완료 했습니다.

## 3 단계 리뷰 반영
- 테스트 네이밍과 관련하여 너무 구체적이지 않게 의도를 나타낼 수 있도록 수정했습니다. 생각을 해보니 테스트 네이밍을 지나치게 구체적으로 적는다면 요구사항의 사소한 변경이 있을 때에도 네이밍 또한 수정 대상이 될 수 있겠다는 생각을 했습니다.

## 4 단계 반영 내용

- 로컬과 테스트 환경을 분리하였습니다. 
- DirtyContext 를 사용하는 이유에 대해서 생각해보고 현재는 인수 테스트 간 데이터 격리 목적으로만 사용하는 것이라고 생각해서 인수테스트 후에 반영된 내용을 모두 지우는 형태로 변경하여 테스트 속도를 향상 시켰습니다.

-  `WholeSecion` : 전체 구간 정보를 가지고 있는 도메인
- `PathGraph` : 현재 저장된 구간 정보로 다익스트라 최단 거리 정보를 가지고 있는 도메인
- `ShortPath` : source to target 의 역 목록과 구간 길이를 가지고 있는 Value Object

처음에는 Line 에서 Sections 을 가지고 있고 모든 Line을 조회해 온뒤에 PathGraph 를 구현했습니다.
하지만 Line 에서 Sections 를 SetUp 하는 과정에서 쿼리가 무겁거나 불필요한 작업 등 Cost 가 많이 발생한다고 생각했고 현재 도메인 요구사항은 모든 Line 에 상관없이 모든 Section 정보만 필요했기에 WholeSecion 을 정의하고 이를 활용하여 PathGraph 를 구현하였습니다.

## 질문 
이전 단계부터 그래왔지만 domain 테스트 , service 테스트  , 인수 테스트 를 모두 수행했습니다.

domain 테스트는  요구사항을 테스트하기 위해 중요한 단위 테스트라고 생각했고
service 테스트는 DB 접근과 통합 하여도 요구사항에 맞게 제대로 동작하는지 실수한 부분이  없는지 확인하기 위해 테스트를 수행했습니다.
인수 테스트는 전체 동작을 테스트하기 위해 중요한 테스트라고 생각합니다.

**하지만 domain 테스트에서 수행했던 내용과 Service 레이어에서 수행했던 테스트 , 인수 테스트 사이클이 굉장히 유사하다고 자주 느꼈고 단 극단적으로는 같은 테스트를 여러번 한다는 생각이 들었습니다.**

이러한 문제에 대해 리뷰어님의 생각이 궁금합니다.

impati committed:
refactor: test displayName 을 조건 - 행위 - 결과를 나타낼 수 있도록 수정

impati committed:
feat: Local 에서 사용할 MySQL 의존성 추가

impati committed:
feat: 데이터베이스 설정을 프로덕션과 테스트에 따라 분리

impati committed:
feat: 프로덕션과 테스트에 따라 분리 재설정

impati committed:
feat: jgrapht 의존성 추가

impati committed:
feat: DirtyContext 로 인수테스트 격리에서 삽입한 데이테베이스 내용을 모두 삭제하는 것으로 변경

impati committed:
test: 경로 조회 기능 인수 테스트 작성

impati committed:
feat: 경로 조회 기능 컨트롤러 구현

impati committed:
feat: 경로를 모두 가져온 뒤 관리하는 WholeSection 정의 및 테스트

impati committed:
feat: JGraph 라이브러리를 이용하여 최단거리 기능 구현 및 테스트

impati committed:
feat: JGraph 라이브러리를 이용하여 최단거리 서비스  구현 및 테스트

impati committed:
refactor: Fixture 및 Helper 개선

impati committed:
refactor: 불필요한 코드 삭제

Laterality commented on src/main/java/subway/application/PathFinder.java:
```diff
@@ -0,0 +1,39 @@
+import subway.dto.response.PathFindResponse;
+
+@Service
+public class PathFinder {
```
서비스의 네이밍 패턴은 일관되게 하면 어떨까요?

impati replied:
좋습니다.!

Laterality commented on src/main/java/subway/domain/WholeSection.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.List;
+import java.util.Set;
+
+public class WholeSection {
```
"전체 구간 정보를 가지고 있는 도메인"이라면 `Sections`가 그 역할을 대신할 수도 있지 않을까요?

기존 도메인을 재활용하는 데 어려움이 있었다고 하셨는데요. 객체 생성 과정은 `Sections`와 `WholeSection`이 같아보이는데, 어떤 부분이 문제였을까요?

impati replied:
##  Sections 의 의미
`Sections` 의 public 기능인 `addSection` ,` deleteSection `모두 같은 `Line`  일때를 가정하고 동작합니다. 
즉 , `Sections` 에서 다른 라인의 구간을 더하게 된다면 제대로 동작하지 않을 가능성이 큽니다. 

예를 들어  `Sections.addSection()` 에는 `throwIfAllOrNothingMatchInLine(newSection)` 로직이 있습니다. 
아예 다른 라인의 구간을 더하고자 할때는 에러를 던지게 됩니다. 
이말은 즉, `Sections` 의 내부 `Section` 들은 암묵적으로 같은 `Line` 의 `Section` 들이어야한다는 의미를 가지고 있습니다.

## 기존 설계 내용
이러한 측면에서 모든 라인의 전체 구간을 담는 기능을 구현하기 위해서 Line 이 내부 필드로 `Sections` 를 가지게 했고 `List<Line>`을 가지는 일급 컬렉션 `Lines` 를 만들어 `Lines` 로 ` PathGraph` 를 구현하고자 했습니다. 
하지만 여기서 `Line` 이 `Sections` 를 가지고 있기위해서 JDBC 에서는 JOIN 으로 모든 값을 채워줘야했고 쿼리 비용이 꽤 많이 든다고 생각했습니다. (더 좋은 방법이 있는데 다소 비효율적이라고 생각했습니다.)

## 그래서 현재 설계
요구사항의 경우 라인에 상관없이 모든 `Section`만 필요했기 때문에 이러한 과정이 의미를 가지기 위해 불필요한 작업을 하고 있다고 생각했고 `findAllSection()` 을 사용하면 요구사항을 만족하기 충분하다고 생각해서 기존 설계를 롤백하고 `findAllSection()` 을 호출하여 모든 `Section` 을 가지고 오는 것으로 설계했습니다.

## 왜  가지고온 모든 구간들을 `Sections` 를 사용하지 않고 별도로 `WholeSection` 을 사용했는지 
앞서 말씀드린대로 `Sections` 는 같은 Line 인 경우 유효합니다.  하지만 가지고 온 모든 구간들은 다른 Line 인 경우도 있기 때문에 이미 `Section` 에서 제공하고 있는 기능들이 유효하게 동작하지 않을 가능성이 매우 컸습니다.(제공하는 기능을 사용하지 않더라도 사용할 수 있게 열려있기 때문에) 또한`PathGraph` 에서는`Sections` 에서 제공하는 기능을 대부분 사용하지 않기 때문에  별도의  모든 라인의 모든 구간을 포함할 수 있다는 의미를 가진 `WholeSection` 를 두어 설계했던 것입니다.

## 정리
1. Sections 는 Line 의 구간들 이라는 의미를 가집니다.
2. 이러한 의미를 표현하고자 Line 에서 Sections 를 필드로 가지고 있는 관계를 가지고자 했습니다.
3. `2` 번의 과정이 현재 요구사항을 해결하기 위해 많이 비효율적이라고 생각했습니다.
4. 현재 요구사항을 효율적으로 만족하고자 모든 노선의 모든 구간을 의미하는 `WholeSection` 만들어 그래프를 구성했습니다.

## 결론
Sections 가 같은 Line 에 속해있는 구간을 의미하는 네이밍으로 변경(LineSections) 하여  사용자가 Sections 의 의미를 네이밍에서 부터 알고자함.

impati replied:
위의 방법으로 해결하는 것이 좋을지는 모르겠습니다.  ORM 기술을 사용했더라면 Line 에서 Sections 를 가지고 있도록 설계했을 것 같습니다.

Laterality reviewed (changes requested):
잘 구현해주셨습니다.

크게 수정할 부분은 없네요. 코멘트 확인해주세요.

## Q&A

"테스트 사이클이 비슷하다"는 것은 "테스트의 내용"이 비슷하다는 말씀일까요, 단순히 "테스트의 실행 주기"가 비슷하다는 말씀일까요?

테스트의 종류별로 목적을 언급하셨는데요. 작성된 테스트가 말씀하신 목적을 달성하기에 충분하지 않은 걸까요?

impati commented:
`"테스트 사이클이 비슷하다"는 것은 "테스트의 내용"이 비슷하다는 말씀일까요, 단순히 "테스트의 실행 주기"가 비슷하다는 말씀일까요?`

**_테스트 내용이 너무 비슷하다_** 는 것을 강조하고 싶었습니다.

개인적으로 중요한 요구사항에 대해 질문에서 말씀드린 3가지 테스트는 반드시 있어야된다고 생각하는데요! 

테스트 내용이 워낙 비슷하다보니 
> 내가 생각하는게 맞나 ?  더 좋은 방법이 있을까 ?

라는 생각이 들어서 질문을 했던 것입니다!

impati committed:
refactor: 네이밍 변경 리팩토링

- PathFinder -> PathService 로 Service Layer 네이밍 통일
- Sections -> LineSections 로 변경
- SortedSections -> SortedLineSections 로 변경

Laterality commented on README.md:
````diff
@@ -57,7 +57,7 @@
 
 ```
-POST /lines/1/sections HTTP/1.1
+POST /lines/1/lineSections HTTP/1.1
````
유효한 경로인가요?

impati replied:
LineSections 로 한번에 바꾸다보니 이런일이 발생한 것 같습니다 수정하겠습니다!

Laterality reviewed (approved):
피드백 잘 반영해주셨습니다 👍 

이번 미션은 이만 머지하도록 하겠습니다.

쉽지 않은 미션이었는데, 고생 많으셨습니다.

## (추가) Q&A

테스트 내용이 비슷한 것은 각 테스트에서 검증하고자 하는 부분들이 많지 않기 때문일 수도 있습니다.

주로 테스트 대상이 다른 클래스의 반환값을 받아 다시 반환하기만 하는 경우 볼 수 있는데요.

그럼에도 각 유형의 테스트가 무의미하지는 않기 때문에, 이런 경우에는 모든 부분을 테스트하려고 하기보다는 필요한 부분을 잘 선정해서 검증하는 편이 테스트도 간결하게 만들고 의도도 알기 쉽습니다.
